js/src/logger.ts

```diff
@@ -1108,123 +1024,98 @@ export async function summarize(
-      "Cannot start a span within a logger from startSpan(). Use logger.startSpan() instead."
-    );
+    return logger;
   }
-
-  return noopSpan;
+  return NOOP_SPAN;
 }
 
 /**
- * Wrapper over `braintrust.startSpan`, which passes the initialized `Span` it to the given callback and ends it afterwards. See `Span.traced` for full details.
+ * Toplevel function for starting a span. It checks the following (in precedence order):
+ *  * Currently-active span
+ *  * Currently-active experiment
+ *  * Currently-active logger
+ *
+ * and creates a span under the first one that is active. If none of these are active, it returns a no-op span object.
+ *
+ * See `Span.traced` for full details.
  */
-export function traced<R>(
+export async function traced<R>(
   callback: (span: Span) => R,
-  args?: StartSpanOptionalNameArgs & SetCurrentArg
-): R {
-  const span = startSpan(args);
-  return runFinally(
-    () => {
-      if (args?.setCurrent ?? true) {
-        return withCurrent(span, () => callback(span));
-      } else {
-        return callback(span);
-      }
-    },
-    () => span.end()
-  );
+  args?: StartSpanArgs & SetCurrentArg
+): Promise<Awaited<R>> {
+  const span = await startSpan(args);
+  try {
+    if (args?.setCurrent ?? true) {
+      return await withCurrent(span, callback);
+    } else {
+      return await callback(span);
+    }
+  } finally {
+    span.end();
+  }
 }
 
 /**
- * Set the given experiment or span as current within the given callback and any asynchronous operations created within the callback. The current experiment can be accessed with `braintrust.currentExperiment`, and the current span with `braintrust.currentSpan`.
- *
- * @param object: The experiment or span to be marked as current.
- * @param callback: The callback to be run under the scope of the current object.
+ * Lower-level alternative to `traced`, which does not automatically end the span or mark it as current. See `traced` for full details.
  */
-export function withCurrent<R>(
-  object: Experiment | Logger | Span,
-  callback: () => R
-): R {
-  if (object.kind === "experiment") {
-    return _state.currentExperiment.run(object, callback);
-  } else if (object.kind === "logger") {
-    return _state.currentLogger.run(object, callback);
-  } else if (object.kind === "span") {
-    return _state.currentSpan.run(object, callback);
-  } else {
-    throw new Error(
-      `Invalid object of type ${(object as any).constructor.name}`
-    );
-  }
+export async function startSpan(args?: StartSpanArgs): Promise<Span> {
+  const parentObject = getSpanParentObject();
+  const { name: nameOpt, ...argsRest } = args ?? {};
+  const name = parentObject.kind === "span" ? nameOpt : nameOpt ?? "root";
+  return parentObject.startSpan({ name, ...argsRest });
+}
+
+// Set the given span as current within the given callback and any asynchronous
+// operations created within the callback.
+function withCurrent<R>(span: Span, callback: (span: Span) => R): R {
```

Comment:
Un-exposing this function is technically a breaking change. If we want to more-softly deprecate it with a warning, we could also do that.

Reply:
Let's just leave it as is